src/libopensc/card-cac.c

```diff
@@ -0,0 +1,1540 @@
+	r = cac_find_first_pki_applet(card, &index);
+	if (r == SC_SUCCESS) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,"applet found, is CAC-1");
+		if (!priv) {
+			priv = cac_new_private_data();
+		}
+		r = cac_populate_cac_1(card, index, priv);
+		if (r == SC_SUCCESS) {
+			card->type = SC_CARD_TYPE_CAC_I;
+			card->drv_data = priv;
+			return r;
+		}
+	}
+	if (priv) {
+		cac_free_private_data(priv);
+	}
+	return r;
+}
+
+
+/* NOTE: returns a bool, 1 card matches, 0 it does not */
+static int cac_match_card(sc_card_t *card)
+{
+	int r;
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	/* Since we send an APDU, the card's logout function may be called...
+	 * however it may be in dirty memory */
+	card->ops->logout = NULL;
+
+	r = cac_find_and_initialize(card);
+	return (r == SC_SUCCESS); /* never match */
+}
+
+
+static int cac_init(sc_card_t *card)
+{
+	int r;
+	unsigned long flags;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	r = cac_find_and_initialize(card);
+	if (r < 0) {
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
+	}
+	flags = SC_ALGORITHM_RSA_RAW;
+
+	_sc_card_add_rsa_alg(card, 1024, flags, 0); /* manditory */
+	_sc_card_add_rsa_alg(card, 2048, flags, 0); /* optional */
+	_sc_card_add_rsa_alg(card, 3072, flags, 0); /* optional */
+
+	card->caps |= SC_CARD_CAP_RNG | SC_CARD_CAP_ISO7816_PIN_INFO;
+
+	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, SC_SUCCESS);
+}
+
+static int cac_pin_cmd(sc_card_t *card, struct sc_pin_cmd_data *data, int *tries_left)
+{
+	/* CAC, like PIV needs Extra validation of (new) PIN during
+	 * a PIN change request, to ensure it's not outside the
+	 * FIPS 201 4.1.6.1 (numeric only) and * FIPS 140-2
+	 * (6 character minimum) requirements.
+	 */
+	struct sc_card_driver *iso_drv = sc_get_iso7816_driver();
+
+	if (data->cmd == SC_PIN_CMD_CHANGE) {
+		int i = 0;
+		if (data->pin2.len < 6) {
+			return SC_ERROR_INVALID_PIN_LENGTH;
+		}
+		for(i=0; i < data->pin2.len; ++i) {
+			if (!isdigit(data->pin2.data[i])) {
+				return SC_ERROR_INVALID_DATA;
+			}
+		}
+	}
+
+	return  iso_drv->ops->pin_cmd(card, data, tries_left);
+}
+
+static struct sc_card_operations cac_ops;
+
+static struct sc_card_driver cac_drv = {
+	"CAC",
+	"cac",
+	&cac_ops,
+	NULL, 0, NULL
+};
+
+static struct sc_card_driver * sc_get_driver(void)
+{
+	struct sc_card_driver *iso_drv = sc_get_iso7816_driver();
+
+	cac_ops = *iso_drv->ops;
+	cac_ops.match_card = cac_match_card;
+	cac_ops.init = cac_init;
+	cac_ops.finish = cac_finish;
+
+	cac_ops.select_file =  cac_select_file; /* need to record object type */
+	cac_ops.get_challenge = cac_get_challenge;
+	cac_ops.read_binary = cac_read_binary;
+	cac_ops.write_binary = cac_write_binary;
+	cac_ops.set_security_env = cac_set_security_env;
+	cac_ops.restore_security_env = cac_restore_security_env;
+	cac_ops.compute_signature = cac_compute_signature;
+	cac_ops.decipher =  cac_decipher;
+	cac_ops.card_ctl = cac_card_ctl;
+	cac_ops.pin_cmd = cac_pin_cmd;
+
+	return &cac_drv;
+}
+
+
+#if 1
```

Comment:
This is present in most of the other card drivers (for some historic reasons?) and therefore it was left here. If it is not needed anymore, I will remove that.